src/builtins/path.cpp

```diff
@@ -0,0 +1,892 @@
+
+        // If we don't have a "." or the "." is the first in the filename,
+        // we do not have an extension
+        size_t pos = filename.find_last_of(L'.');
+        if (pos == wcstring::npos || pos == 0) {
+            return none();
+        }
+
+        // Convert pos back to what it would be in the original path.
+        return pos + path.size() - filename.size();
+}
+
+static int path_extension(parser_t &parser, io_streams_t &streams, int argc, const wchar_t **argv) {
+    options_t opts;
+    int optind;
+    int retval = parse_opts(&opts, &optind, 0, argc, argv, parser, streams);
+    if (retval != STATUS_CMD_OK) return retval;
+
+    int n_transformed = 0;
+    arg_iterator_t aiter(argv, optind, streams, opts.null_in);
+    while (const wcstring *arg = aiter.nextstr()) {
+        auto pos = find_extension(*arg);
+
+        if (!pos) {
+            // If there is no extension the extension is empty.
+            // This is unambiguous because we include the ".".
+            path_out(streams, opts, L"");
+            continue;
+        }
+
+        wcstring ext = arg->substr(*pos);
+        if (opts.quiet && !ext.empty()) {
+            return STATUS_CMD_OK;
+        }
+        path_out(streams, opts, ext);
+        n_transformed++;
+    }
+
+    return n_transformed > 0 ? STATUS_CMD_OK : STATUS_CMD_ERROR;
+}
+
+static int path_change_extension(parser_t &parser, io_streams_t &streams, int argc, const wchar_t **argv) {
+    options_t opts;
+    int optind;
+    int retval = parse_opts(&opts, &optind, 1, argc, argv, parser, streams);
+    if (retval != STATUS_CMD_OK) return retval;
+
+    int n_transformed = 0;
+    arg_iterator_t aiter(argv, optind, streams, opts.null_in);
+    while (const wcstring *arg = aiter.nextstr()) {
+        auto pos = find_extension(*arg);
+
+        wcstring ext;
+        if (!pos) {
+            ext = *arg;
+        } else {
+            ext = arg->substr(0, *pos);
+        }
+
+        // Only add on the extension "." if we have something.
+        // That way specifying an empty extension strips it.
+        if (*opts.arg1) {
+            if (opts.arg1[0] != L'.') {
+                ext.push_back(L'.');
+            }
+            ext.append(opts.arg1);
+        }
+        path_out(streams, opts, ext);
+        n_transformed++;
+    }
+
+    return n_transformed > 0 ? STATUS_CMD_OK : STATUS_CMD_ERROR;
+}
+
+static int path_resolve(parser_t &parser, io_streams_t &streams, int argc, const wchar_t **argv) {
+    options_t opts;
+    int optind;
+    int retval = parse_opts(&opts, &optind, 0, argc, argv, parser, streams);
+    if (retval != STATUS_CMD_OK) return retval;
+
+    int n_transformed = 0;
+    arg_iterator_t aiter(argv, optind, streams, opts.null_in);
+    while (const wcstring *arg = aiter.nextstr()) {
+        auto real = wrealpath(*arg);
+
+        if (!real) {
+            // The path doesn't exist, so we go up until we find
```

Comment:
There's three important reasons for failure here:

1. `ENOENT/ENOTDIR`: Path does not exist
2. `EPERM`: Path may exist but you can't read it
3. `ELOOP`: Path is a symlink loop

For a symlink loop, it probably makes sense to "resolve as far as we can" as this does, but it would be good to add a test for that case since it's easily overlooked. Also we should change the language in the docs from "nonexistent paths" to "paths which do not exist, or otherwise cannot be resolved" or something, maybe you have ideas here.